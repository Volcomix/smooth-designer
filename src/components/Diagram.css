.Diagram {
  flex: 1;
  background-color: var(--grey-50);
}

.Diagram-content {
  width: 0;
  height: 0;
  position: absolute;
  top: 50%;
  left: 50%;
}

.Diagram-add {
  position: fixed;
  bottom: 24px;
  right: 24px;
  z-index: 100;
}
