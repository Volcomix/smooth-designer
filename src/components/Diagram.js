//@flow
import React from 'react'
import FloatingActionButton from 'material-ui/FloatingActionButton'
import ContentAdd from 'material-ui/svg-icons/content/add'
import './Diagram.css'
import type { Block } from '../types'
import MousePosition from './MousePosition'
import BlockDetail from './BlockDetail'
import LinkDetail from './LinkDetail'
import Linking from './Linking'

export type Props = {
  blocks: Block[],
  links: { id: string, fromBlock: Block, toBlock: Block }[],
  linking: ?{ fromBlock: Block, toMouse: { x: number, y: number } },
  onAddClick: () => void,
  onBlockDelete: (id: string) => void,
  onNameChange: (id: string, name: string) => void,
  onSizeChange: (id: string, width: number, height: number) => void,
  onLinkStart: (id: string, mouseX: number, mouseY: number) => void,
  onLinkMove: (mouseX: number, mouseY: number) => void,
  onLinkEnd: (id: string) => void,
  onLinkCancel: () => void,
  onLinkDelete: (id: string) => void,
}

const Diagram = ({
  blocks,
  links,
  linking,
  onAddClick,
  onBlockDelete,
  onNameChange,
  onSizeChange,
  onLinkStart,
  onLinkMove,
  onLinkEnd,
  onLinkCancel,
  onLinkDelete,
}: Props) => {
  const isLinking = !!linking
  return (
    <MousePosition
      className={'Diagram' + (isLinking ? ' Diagram-linking' : '')}
      isActive={isLinking}
      onMouseMove={onLinkMove}
      onMouseUp={() => {
        document.activeElement && document.activeElement.blur()
        onLinkCancel()
      }}
    >
      <div className="Diagram-content">
        {blocks.map(block => (
          <BlockDetail
            key={block.id}
            {...block}
            isLinking={isLinking}
            onDelete={onBlockDelete}
            onNameChange={onNameChange}
            onSizeChange={onSizeChange}
            onLinkStart={onLinkStart}
            onLinkEnd={onLinkEnd}
          />
        ))}
        {links.map(({ id, fromBlock, toBlock }) => (
          <LinkDetail
            key={id}
            id={id}
            fromBlock={fromBlock}
            toBlock={toBlock}
            onDelete={onLinkDelete}
          />
        ))}
        {linking && <Linking {...linking} />}
      </div>
      <FloatingActionButton
        className="Diagram-add"
        secondary={true}
        zDepth={4}
        onClick={onAddClick}
      >
        <ContentAdd />
      </FloatingActionButton>
    </MousePosition>
  )
}

export default Diagram
