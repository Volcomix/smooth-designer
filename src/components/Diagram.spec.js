//@flow
import React from 'react'
import { shallow } from 'enzyme'
import Diagram, { type Props } from './Diagram'

const setup = setupProps => {
  const defaultProps: Props = {
    blocks: [],
    links: [],
    linking: undefined,
    onAddClick: jest.fn(),
    onBlockDelete: jest.fn(),
    onNameChange: jest.fn(),
    onSizeChange: jest.fn(),
    onLinkStart: jest.fn(),
    onLinkMove: jest.fn(),
    onLinkEnd: jest.fn(),
    onLinkCancel: jest.fn(),
    onLinkDelete: jest.fn(),
  }
  const props = { ...defaultProps, ...setupProps }
  const wrapper = shallow(<Diagram {...props} />)
  return { wrapper, props }
}

it('renders without crashing', () => {
  const { wrapper } = setup()
  expect(wrapper).toMatchSnapshot()
})

it('calls onAddClick when the Add button is clicked', () => {
  const { wrapper, props } = setup()
  wrapper.find('.Diagram-add').simulate('click')
  expect(props.onAddClick).toHaveBeenCalled()
})

describe('with blocks', () => {
  it('renders the blocks', () => {
    const { wrapper } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 10, height: 20 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 30, height: 40 },
      ],
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a link if linking', () => {
    const { wrapper } = setup({
      blocks: [{ id: '0', name: 'Block', x: 0, y: 0, width: 10, height: 20 }],
      linking: {
        fromBlock: {
          id: '0',
          name: 'Block',
          x: 1,
          y: 2,
          width: 0,
          height: 0,
        },
        toMouse: { x: 10, y: 20 },
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('calls onBlockDelete when a block is deleted', () => {
    const { wrapper, props } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 0, height: 0 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 0, height: 0 },
      ],
    })
    wrapper
      .find('BlockDetail')
      .first()
      .simulate('delete', '0')
    expect(props.onBlockDelete).toHaveBeenCalledWith('0')
  })

  it('calls onNameChange when a block name changes', () => {
    const { wrapper, props } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 0, height: 0 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 0, height: 0 },
      ],
    })
    wrapper
      .find('BlockDetail')
      .first()
      .simulate('nameChange', '0', 'New name')
    expect(props.onNameChange).toHaveBeenCalledWith('0', 'New name')
  })

  it('calls onSizeChange when a block size changes', () => {
    const { wrapper, props } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 0, height: 0 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 0, height: 0 },
      ],
    })
    wrapper
      .find('BlockDetail')
      .first()
      .simulate('sizeChange', '0', 10, 20)
    expect(props.onSizeChange).toHaveBeenCalledWith('0', 10, 20)
  })

  it('calls onLinkStart when linking starts', () => {
    const { wrapper, props } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 0, height: 0 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 0, height: 0 },
      ],
    })
    wrapper
      .find('BlockDetail')
      .first()
      .simulate('linkStart', '0', 10, 20)
    expect(props.onLinkStart).toHaveBeenCalledWith('0', 10, 20)
  })

  it('calls onLinkMove when linking moves', () => {
    const { wrapper, props } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 0, height: 0 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 0, height: 0 },
      ],
      linking: {
        fromBlock: {
          id: '0',
          name: 'Block 1',
          x: 0,
          y: 0,
          width: 0,
          height: 0,
        },
        toMouse: { x: 10, y: 20 },
      },
    })
    wrapper.simulate('mouseMove', 30, 40)
    expect(props.onLinkMove).toHaveBeenCalledWith(30, 40)
  })

  it('calls onLinkEnd when linking ends', () => {
    const { wrapper, props } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 0, height: 0 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 0, height: 0 },
      ],
      linking: {
        fromBlock: {
          id: '1',
          name: 'Block 1',
          x: 0,
          y: 0,
          width: 0,
          height: 0,
        },
        toMouse: { x: 10, y: 20 },
      },
    })
    wrapper
      .find('BlockDetail')
      .first()
      .simulate('linkEnd', '0')
    expect(props.onLinkEnd).toHaveBeenCalledWith('0')
  })

  it('calls onLinkCancel when the mouse is up', () => {
    const { wrapper, props } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 0, height: 0 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 0, height: 0 },
      ],
      linking: {
        fromBlock: {
          id: '1',
          name: 'Block 1',
          x: 0,
          y: 0,
          width: 0,
          height: 0,
        },
        toMouse: { x: 10, y: 20 },
      },
    })
    const spy = jest.spyOn(document.activeElement || {}, 'blur')
    wrapper.simulate('mouseUp')
    expect(spy).toHaveBeenCalled()
    expect(props.onLinkCancel).toHaveBeenCalled()
  })
})

describe('with links', () => {
  it('renders the links', () => {
    const { wrapper } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 10, height: 20 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 30, height: 40 },
        { id: '2', name: 'Block 3', x: 0, y: 0, width: 50, height: 60 },
      ],
      links: [
        {
          id: '0',
          fromBlock: {
            id: '0',
            name: 'Block 1',
            x: 0,
            y: 0,
            width: 10,
            height: 20,
          },
          toBlock: {
            id: '1',
            name: 'Block 2',
            x: 0,
            y: 0,
            width: 30,
            height: 40,
          },
        },
        {
          id: '1',
          fromBlock: {
            id: '1',
            name: 'Block 2',
            x: 0,
            y: 0,
            width: 30,
            height: 40,
          },
          toBlock: {
            id: '2',
            name: 'Block 3',
            x: 0,
            y: 0,
            width: 50,
            height: 60,
          },
        },
      ],
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('calls onLinkDelete when a link is deleted', () => {
    const { wrapper, props } = setup({
      blocks: [
        { id: '0', name: 'Block 1', x: 0, y: 0, width: 0, height: 0 },
        { id: '1', name: 'Block 2', x: 0, y: 0, width: 0, height: 0 },
        { id: '2', name: 'Block 3', x: 0, y: 0, width: 0, height: 0 },
      ],
      links: [
        {
          id: '0',
          fromBlock: {
            id: '0',
            name: 'Block 1',
            x: 0,
            y: 0,
            width: 0,
            height: 0,
          },
          toBlock: {
            id: '1',
            name: 'Block 2',
            x: 0,
            y: 0,
            width: 0,
            height: 0,
          },
        },
        {
          id: '1',
          fromBlock: {
            id: '1',
            name: 'Block 2',
            x: 0,
            y: 0,
            width: 0,
            height: 0,
          },
          toBlock: {
            id: '2',
            name: 'Block 3',
            x: 0,
            y: 0,
            width: 0,
            height: 0,
          },
        },
      ],
    })
    wrapper
      .find('LinkDetail')
      .first()
      .simulate('delete', '0')
    expect(props.onLinkDelete).toHaveBeenCalledWith('0')
  })
})
