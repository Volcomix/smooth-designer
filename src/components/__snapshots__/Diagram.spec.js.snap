// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<MousePosition
  className="Diagram"
  isActive={false}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="Diagram-content"
  />
  <FloatingActionButton
    className="Diagram-add"
    disabled={false}
    mini={false}
    onClick={[Function]}
    secondary={true}
    zDepth={4}
  >
    <ContentAdd />
  </FloatingActionButton>
</MousePosition>
`;

exports[`with blocks renders a link if linking 1`] = `
<MousePosition
  className="Diagram Diagram-linking"
  isActive={true}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="Diagram-content"
  >
    <BlockDetail
      height={0}
      id="0"
      isLinking={true}
      key="0"
      name="Block"
      onDelete={[Function]}
      onLinkEnd={[Function]}
      onLinkStart={[Function]}
      onNameChange={[Function]}
      onSizeChange={[Function]}
      width={0}
      x={0}
      y={0}
    />
    <Linking
      fromBlock={
        Object {
          "height": 0,
          "id": "0",
          "name": "Block",
          "width": 0,
          "x": 1,
          "y": 2,
        }
      }
      toMouse={
        Object {
          "x": 10,
          "y": 20,
        }
      }
    />
  </div>
  <FloatingActionButton
    className="Diagram-add"
    disabled={false}
    mini={false}
    onClick={[Function]}
    secondary={true}
    zDepth={4}
  >
    <ContentAdd />
  </FloatingActionButton>
</MousePosition>
`;

exports[`with blocks renders the blocks 1`] = `
<MousePosition
  className="Diagram"
  isActive={false}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="Diagram-content"
  >
    <BlockDetail
      height={0}
      id="0"
      isLinking={false}
      key="0"
      name="Block 1"
      onDelete={[Function]}
      onLinkEnd={[Function]}
      onLinkStart={[Function]}
      onNameChange={[Function]}
      onSizeChange={[Function]}
      width={0}
      x={0}
      y={0}
    />
    <BlockDetail
      height={0}
      id="1"
      isLinking={false}
      key="1"
      name="Block 2"
      onDelete={[Function]}
      onLinkEnd={[Function]}
      onLinkStart={[Function]}
      onNameChange={[Function]}
      onSizeChange={[Function]}
      width={0}
      x={0}
      y={0}
    />
  </div>
  <FloatingActionButton
    className="Diagram-add"
    disabled={false}
    mini={false}
    onClick={[Function]}
    secondary={true}
    zDepth={4}
  >
    <ContentAdd />
  </FloatingActionButton>
</MousePosition>
`;

exports[`with links renders the links 1`] = `
<MousePosition
  className="Diagram"
  isActive={false}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="Diagram-content"
  >
    <BlockDetail
      height={0}
      id="0"
      isLinking={false}
      key="0"
      name="Block 1"
      onDelete={[Function]}
      onLinkEnd={[Function]}
      onLinkStart={[Function]}
      onNameChange={[Function]}
      onSizeChange={[Function]}
      width={0}
      x={0}
      y={0}
    />
    <BlockDetail
      height={0}
      id="1"
      isLinking={false}
      key="1"
      name="Block 2"
      onDelete={[Function]}
      onLinkEnd={[Function]}
      onLinkStart={[Function]}
      onNameChange={[Function]}
      onSizeChange={[Function]}
      width={0}
      x={0}
      y={0}
    />
    <BlockDetail
      height={0}
      id="2"
      isLinking={false}
      key="2"
      name="Block 3"
      onDelete={[Function]}
      onLinkEnd={[Function]}
      onLinkStart={[Function]}
      onNameChange={[Function]}
      onSizeChange={[Function]}
      width={0}
      x={0}
      y={0}
    />
    <LinkDetail
      fromBlock={
        Object {
          "height": 0,
          "id": "0",
          "name": "Block 1",
          "width": 0,
          "x": 0,
          "y": 0,
        }
      }
      id="0"
      key="0"
      onDelete={[Function]}
      toBlock={
        Object {
          "height": 0,
          "id": "1",
          "name": "Block 2",
          "width": 0,
          "x": 0,
          "y": 0,
        }
      }
    />
    <LinkDetail
      fromBlock={
        Object {
          "height": 0,
          "id": "1",
          "name": "Block 2",
          "width": 0,
          "x": 0,
          "y": 0,
        }
      }
      id="1"
      key="1"
      onDelete={[Function]}
      toBlock={
        Object {
          "height": 0,
          "id": "2",
          "name": "Block 3",
          "width": 0,
          "x": 0,
          "y": 0,
        }
      }
    />
  </div>
  <FloatingActionButton
    className="Diagram-add"
    disabled={false}
    mini={false}
    onClick={[Function]}
    secondary={true}
    zDepth={4}
  >
    <ContentAdd />
  </FloatingActionButton>
</MousePosition>
`;
